csrc/scheduler/normalization_inner.cpp

```diff
@@ -252,6 +263,335 @@ int64_t getMaxPersistentBatch(
+struct HeuristicParams {
+  int64_t bdimx = -1;
+  int64_t bdimy = -1;
+  int64_t padded_bdimx = -1;
+  int64_t persistent_batch_size = -1;
+  int64_t register_per_thread = -1;
+  int64_t free_registers = -1;
+  int64_t occupancy = -1;
+  int64_t n_persistent_tails = -1;
+  bool is_pad_bdimx = false;
+  void print() const {
+    std::cout << "bdimx: " << bdimx << ", bdimy: " << bdimy
+              << ", padded_bdimx: " << padded_bdimx
+              << ", persistent_batch_size: " << persistent_batch_size
+              << ", register_per_thread: " << register_per_thread
+              << ", free_registers: " << free_registers
+              << ", occupancy: " << occupancy
+              << ", n_persistent_tails: " << n_persistent_tails
+              << ", is_pad_bdimx: " << is_pad_bdimx << std::endl;
+  }
+};
+
+HeuristicParams getHeuristicParamsGivenPerisisentBatchSize(
+    const int64_t reduction_count_after_vectorize,
+    const int64_t total_iteration_numel,
+    const int64_t max_multi_reduction_factor,
+    const int64_t min_threads_per_block,
+    const int64_t buffer_bytes_per_batch,
+    const int64_t target_warps_per_sm,
+    const int64_t free_registers,
+    const int64_t persistent_batch_size) {
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  auto device_warp_size = dev_prop->warpSize;
+  auto max_threads_per_block = dev_prop->maxThreadsPerBlock;
+  HeuristicParams params;
+  params.persistent_batch_size = persistent_batch_size;
+
+  // set bdimx and bdimy
+  params.bdimx = scheduler_utils::safeDiv(
+      reduction_count_after_vectorize, persistent_batch_size);
+  NVF_ERROR(
+      params.bdimx <= max_threads_per_block,
+      "persistent batch size too small! bdimx should be less than ",
+      max_threads_per_block,
+      ", but got ",
+      params.bdimx);
+  params.bdimy = std::min(
+      scheduler_utils::safeDiv(min_threads_per_block, params.bdimx),
+      max_multi_reduction_factor);
+  params.padded_bdimx = params.bdimx % device_warp_size == 0
+      ? params.bdimx
+      : params.bdimx + (device_warp_size - params.bdimx % device_warp_size);
+  params.is_pad_bdimx = params.bdimx > 16 &&
+      params.padded_bdimx * params.bdimy <= max_threads_per_block;
+
+  // calculate register per thread and achieved occupancy
+  int64_t threads_per_block = params.is_pad_bdimx
+      ? params.padded_bdimx * params.bdimy
+      : params.bdimx * params.bdimy;
+  int64_t persistent_buffer_size =
+      buffer_bytes_per_batch * persistent_batch_size;
+  auto reg_occ = getMaxRegisterCountPerThreadAndOccupancy(
+      persistent_buffer_size,
+      threads_per_block,
+      target_warps_per_sm,
+      free_registers);
+  params.register_per_thread = reg_occ.first;
+  params.occupancy = reg_occ.second;
+  params.free_registers = params.register_per_thread -
+      persistent_buffer_size / scheduler_utils::bytes_per_register;
+  // (4) Calculate other quantities reflecting the quality of the heuristic.
```

Comment:
nitpick, what the `(4)` here?

Reply:
refering to `n_persistent_tails`, it is not a heuristic para but it is an indicator of the quality of the heuristic. In other words, it is a para used in the sorting.